elephant/phase_analysis.py

```diff
@@ -307,3 +308,92 @@ def phase_difference(alpha, beta):
     delta = alpha - beta
     phase_diff = np.arctan2(np.sin(delta), np.cos(delta))
     return phase_diff
+
+
+def weighted_phase_lag_index(signal_i, signal_j, sampling_frequency=None,
+                             absolute_value=True):
+    r"""
+    Calculates the Weigthed Phase-Lag Index (WPLI)
+
+    This function estimates the WPLI, which is a measure of phase-synchrony. It
+    describes for two given signals i and j, who is leading/lagging the other
+    signal in the frequency domain across multiple trials.
+
+    Parameters
+    ----------
+    signal_i, signal_j : np.array, Quantity, neo.AnalogSignal
+        Time-series of the first and second signals,
+        with `t` time points and `n` trials.
+    sampling_frequency : quantity (default: None)
+        Sampling frequency of the signals in Hz. Not needed if signal i and j
+        are neo.AnalogSignals.
+    absolute_value : boolean (default: True)
+        Takes the absolute value of the numerator in the WPLI-formula.
+        When set to `False`, the WPLI contains additional directionality
+        information about which signal leads/lags the other signal:
+            - wpli > 0 : first signal leads second one
+            - wpli < 0 : first signal lags second one
+
+    Returns
+    -------
+    freqs : quantity
+        Positive frequencies in Hz associated with the estimates of `wpli`.
+        Range: :math:`[0, sampling frequency/2]`
+    wpli : float
+        Weighted phase-lag index of `signal_i` and `signal_j` across trials.
+        Range: :math:`[0, 1]`
+
+    Raises
+    ------
+    ValueError
+        If trial number or trial length are different for signal i and j.
+
+    Notes
+    -----
+    This implementation is based on the formula taken from [1] (pp.1550) :
+
+    .. math::
+        WPLI = \frac{| E( |Im(X)| * sgn(Im(X)) ) |}{E( |Im(X)| )}
+
+    with:
+        - E{...} : expected value operator
+        - Im{X} : imaginary component of the cross-spectrum
+        - X = Z_i * Z_j_conjugate : as cross-spectrum
+        - Z_i, Z_j : complex-valued matrix, representing the Fourier spectra
+                     of a particular frequency of the signals i and j.
+
+    References
+    ----------
+    ..  [1] Martin Vinck, Robert Oostenveld, Marijn van Wingerden,
+        Franscesco Battaglia, Cyriel M.A. Pennartz; "An improved index of
+        phase-synchronization for electrophysiological data in the presence
+        of volume-conduction, noise and sample-size bias"; NeuroImage, vol. 55,
+        pp. 1548-1565, 2011
+
+    """
+    if sampling_frequency is None:  # neo.AnalogSignal-input
```

Comment:
It will be safer if checking the input types to be `neo.AnalogSignal` when None, otherwise this will raise an unhandled exception if the user passes a NumPy array and the rate is None. It should also raise an exception if not passing the sample rate when needed. Moreover, what if the sampling rates in `signal_i` and `signal_j` are different?

Reply:
new commit 39785fe7cb60d2b5a829b5a5ae4f391ef0c8917a is regarding this now